update/3329: Changed autocomplete menu "ul > li > button" to "div > button"

BoardJames opened the pull request:
## Description
As requested in issue #3329 I have changed the structure for the autocomplete menu to be a div with buttons. This will hopefully improve how screen readers respond so they note that the button is clickable though I did not notice any changed in the behaviour of NVDA.
<!-- Please describe your changes -->

## How Has This Been Tested?
Existing automated tests cover the autocomplete menu. I have also manually verified the changes in Firefox with NVDA.
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots (jpeg or gifs if applicable):

## Types of changes
Minor changes to structure and CSS of the autocomplete menu which might be considered a bug fix.
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

tiny-james committed:
update/3329: Changed autocomplete menu "ul > li > button" to "div > button"

This will hopefully improve how screen readers respond so they note that
the button is clickable.

tiny-james committed:
update/3329: Updated selector to be more specific to fix tests

afercia commented:
From an a11y perspective looks good to me, thanks! Tested with Safari+VoiceOver and Firefox+NVDA. The last combo is also able to announce the item number within the set of results.

It would be great to add a `speak` message to announce the total number of results, as done for the link and tags suggestions. Something like `nn results found, use up and down arrow keys to navigate.` Please let me know if you want me to open a separate issue.

I see a small visual glitch, see last screenshot. When username/first name/last name are very long and depending also on the option `Display name publicly as`, the text in a suggestion can be very long. In Chrome, the text can overflow from the "popover". In Firefox, the images shrink: I guess they need `flex: 0 0 auto;`.

![screen shot 2017-11-06 at 14 19 34](https://user-images.githubusercontent.com/1682452/32444416-d45cdd24-c302-11e7-92c6-2b305667386c.png)

![screenshot 101](https://user-images.githubusercontent.com/1682452/32444415-d43c2f2a-c302-11e7-973f-b482647790ce.png)

Images with very long suggestions text in Firefox:

![screen shot 2017-11-06 at 14 24 34](https://user-images.githubusercontent.com/1682452/32444932-886f6c54-c304-11e7-94f4-4921aba2e7a7.png)

Also, In Safari, the after selecting an user from the suggestion (either by clicking or via the keyboard) the caret position is at the beginning of the inserted user name, before the "at" sign. I think it's unrelated to this PR though (boundaries?).

Safari caret position after inserting an username:

![screen shot 2017-11-06 at 14 21 39](https://user-images.githubusercontent.com/1682452/32445080-ffda8198-c304-11e7-8645-5dce2e35b230.png)

Question: from a functional point of view, what happens when there are hundreds/thousands of users on a site? What is the max number of users returned by the API? Would the API need some adjustments?

afercia reviewed (changes requested):
I think at least the CSS issue should be fixed in this PR. I guess the other points would need their own separate issues.

EphoxJames commented:
Hmmm, looks like I need to set flex-shrink to 0. 
> Question: from a functional point of view, what happens when there are hundreds/thousands of users on a site? What is the max number of users returned by the API? Would the API need some adjustments?

I don't know anything about the collections API but the autocomplete popup only shows 10 items at a time so you would need to filter them by typing.

tiny-james committed:
update/3329: Fix styling - images don't shrink, text has max-width

tiny-james committed:
Merge branch 'master' into update/3329-autocomplete-aria-markup-improvement

EphoxJames commented:
I have made it handle long names.
![selection_058](https://user-images.githubusercontent.com/26243021/32471493-4f85b042-c3a9-11e7-896b-6feca2f5ab49.png)

The profile image will always stay the same size (24px x 24px). If the display name is longer than 200px it will get truncated and if the slug is longer than 100px it will be truncated. Both the display name and the slug have `white-space: nowrap` set.

tiny-james committed:
update/3329: specify profile picture size to avoid resizing after load

tiny-james committed:
update/3329: ESC suppresses popup, CTRL+SPACE reshows; List count spoken;

Filtering is now done on typing and load instead of on rendering;
List item id is now tied to key instead of index so it will change on
filtering which will hopefully lead to screen readers noticing the
change more often

EphoxJames commented:
I've added a few more changes.

1. The list count will now be spoken just like the tags input
2. Pressing ESC will now suppress the popup (not just hide it until a new character is typed)
3. Pressing CTRL+SPACE while the popup is suppressed will show it again.
4. The ID assigned to the list items should now be guaranteed to be different for different list items whereas before it depended on the position in the filtered list which meant that it could stay the same when the options changed due to filtering
5. Filtering is done on typing instead of render which I think will be more efficient (and it made implementing the spoken count easier).

tiny-james committed:
Merge branch 'master' into update/3329-autocomplete-aria-markup-improvement

EphoxJames commented:
I haven't solved the Safai cursor position problem yet - in theory it shouldn't be happing because the link is inserted before the `@query` and then the `@query` is removed so there should be no way for the cursor to end up before the link. I will probably have to set the selection to fix it.

afercia commented:
> The list count will now be spoken just like the tags input

Thanks! The results are nicely announced, tested with Safari+VoiceOver and Firefox+NVDA.

> Pressing CTRL+SPACE while the popup is suppressed will show it again.

This doesn't seem to work for me. When previously pressing Escape, the block is not focused any longer so CTRL+SPACE doesn't do anything for me.

From an a11y perspective, everything seems to work fine to me. I'd leave the overall technical review to someone with more technical expertise than me :)

I'd suggest to consider just one small usability improvement: on Twitter, typing `@` doesn't do anything and the suggestions list is displayed only after typing at least one character after the `@`. I'd consider to do the same here.

afercia commented:
Note: on Safari 11.0.1 and only when using VoiceOver, the boundaries don't work at all for me: no highlighting is displayed on both user names and links. No "link modal" appears. They work on Safari Technology Preview – Release 43 (Safari 11.1, WebKit 12605.1.12). Seems to me something that should be monitored in a separate issue.a

EphoxJames commented:
> I'd suggest to consider just one small usability improvement: on Twitter, typing @ doesn't do anything and the suggestions list is displayed only after typing at least one character after the @. I'd consider to do the same here.

Both Facebook and Slack display a list after only typing the '@'. I tried it anyway since it was technically a very easy change and it just feels wrong to not have the pop-up appear so I'm going to not implement that.

Note if you want to try it out then modify line 284-286 of components/autocomplete/index.js to:
```
			if ( text.substr( pos + 1, triggerPrefix.length ) !== triggerPrefix ||
					text.length <= ( pos + 1 + triggerPrefix.length ) ) {
				return false;
			}
```

afercia commented:
Thanks! Fair enough for the first character :) I don't have a strong preference, to me the Twitter behavior seems more intuitive but since it's an easy change, can also be evaluated in the future.

Two more considerations:
Should the "search term" match usernames that contain the search term or usernames that starts with it? At the moment, seems to me it uses "contain", see below searching for "me":

![screen shot 2017-11-08 at 08 58 15](https://user-images.githubusercontent.com/1682452/32537908-806d3f52-c464-11e7-816a-9d2596b82b2d.png)

it returns usernames containing "me". Instead, seems to me both Slack and Twitter search for matches that _start_ with the search term. As far as I see, they search for name or last name or "slug" starting with the typed search term. Thoughts? /cc @jasmussen 

Minor:
I'd consideer to reset the border radius on the focused search result: all similar menus in Gutenberg (e.g. block switcher menu, edit table menu, use squared borders.

jasmussen commented:
I'm general for the auto complete stuff, the behavior I'd mimic is that of IRC and slack when you type /. So yes, more like auto complete, less of a search. Nice work though.

@iseulde should also be heard on the syntactic sugar like how to start quotes, lists and headings. She has a ton of experience on how that works best.

Regarding the border radii, that's a regression in master. Okay to fix, okay to ignore also, as I'm working separately on improving all those focus styles in a separate pr.

tiny-james committed:
Merge branch 'master' into update/3329-autocomplete-aria-markup-improvement

tiny-james committed:
update/3329: Autocompleter matches after word boundaries or space

EphoxJames commented:
@jasmussen I have changed it to only start matches on word boundaries or after spaces so it won't match things in the middle of words any more.

jasmussen commented:
Nice, that works well! I'd guess this is ready for final code review. Nice work!

afercia commented:
Tested again also on Windows and from an accessibility perspective, it works pretty well. Thanks!

Unfortunately, with IE 11 nothing happens 😞 no suggestions list (seems to me it's not rendered in the DOM), no speak message, nothing at all.

IE 11:

![screenshot 103](https://user-images.githubusercontent.com/1682452/32605028-1cd1b3a0-c550-11e7-9744-8c48ec2c463b.png)

For reference: Firefox and NVDA announce everything pretty well:

![screenshot 102](https://user-images.githubusercontent.com/1682452/32605052-345ae118-c550-11e7-9d86-c56c3663666c.png)

aduth commented on components/autocomplete/index.js:
```diff
@@ -247,88 +299,67 @@ export class Autocomplete extends Component {
-		if ( options.length === 0 ) {
+		if ( suppress === open.idx ) {
+			const { keyCode, ctrlKey, shiftKey, altKey, metaKey } = event;
+			if ( keyCode === SPACE && ctrlKey && ! ( shiftKey || altKey || metaKey ) ) {
```
I'm wondering about the `SPACE` check here; if the user moves focus out of the block after dismissing suggestions and then returns later, shouldn't we expect the dismissal to no longer take effect?

![focus-out](https://user-images.githubusercontent.com/1779930/32625153-800e9eca-c559-11e7-80dd-b7c2a25b6400.gif)

aduth replied:
In a future refactor, to simplify the handling of modifier keys, we might consider adding support to `<KeyboardShortcuts />` to monitor key events of children, mentioned originally in implementation notes #1944 . Made more difficult here by the fact that we manually create the DOM event handlers.

EphoxJames replied:
Yep, that still hidden after move thing is a bug.

EphoxJames replied:
The bug should be fixed.

mcsf commented on blocks/autocompleters/style.scss:
```diff
@@ -2,11 +2,34 @@
+		width: 24px; // avoid jarring resize by seting the size upfront
+		height: 24px;
+	}
+	.name {
```
Although they weren't introduced in this PR, these classnames don't adhere to Gutenberg's [BEM-like conventions](https://github.com/WordPress/gutenberg/blob/a0714e2d85e9c985a9059b45e7f984d3fc780e63/docs/coding-guidelines.md#naming). We should have something along the lines of `components-autocomplete__result-name`.

EphoxJames replied:
Hmmm, reading that document I'd come up with `blocks-autocompleters__user-name` because the actual CSS is not in the `components/autocomplete` folder but the `blocks/autocompleters` folder. @mcsf  Does that sound right to you?
I would also have `blocks-autocompleters__user-slug` and `blocks-autocompleters__user-avatar`.

EphoxJames replied:
I have renamed the classes. I also tweaked where the completers' className is applied in the Autocompleter structure so their css won't care about the Autocompleter's class names.

mcsf commented on components/autocomplete/index.js:
```diff
@@ -247,88 +299,67 @@ export class Autocomplete extends Component {
+		// filter the options we already have
+		const filteredOptions = open ? filterOptions( search, this.state[ 'options_' + open.idx ] ) : [];
+		// check if we should still suppress the popover
+		const suppress = open && wasSuppress === open.idx ? wasSuppress : undefined;
```
I suggest adding parentheses to make eval order really obvious, instead of relying on precedence rules. One level is enough: `( a && b === c ) ? d : e`

EphoxJames replied:
Done

tiny-james committed:
Merge branch 'master' into update/3329-autocomplete-aria-markup-improvement

tiny-james committed:
update/3329: Fix the list staying suppressed when the cursor is moved

EphoxJames commented:
@aduth It seems that your new test "closes by blur" doesn't like my changes in this branch but I can't figure out why. Any ideas?

tiny-james committed:
Merge branch 'master' into update/3329-autocomplete-aria-markup-improvement

tiny-james committed:
update/3329: fixed withFocusOutside position in call order

tiny-james committed:
update/3329: Renamed classes to follow BEM style guide

tiny-james committed:
update/3329: Bracketed test of turnary statement to improve readability

EphoxJames commented:
I figured out that the "closes by blur" test was failing because I had put the withSpokenMessages HOC before the withFocusOutside in the evaluation order and withFocusOutside was depending on calling the  handleFocusOutside method of the child component (which withFocusOutside did not have).

EphoxJames commented:
I think I have addressed all review feedback so I will merge.

aduth commented on blocks/autocompleters/style.scss:
```diff
@@ -1,12 +1,35 @@
-	color: $dark-gray-100;
-}
+.blocks-autocompleters__user {
+	.blocks-autocompleters__user-avatar {
```
One of the benefits of using a BEM convention is that it eliminates the need for nesting in many cases, so this could be moved to a top-level style definition.

aduth commented on components/autocomplete/index.js:
```diff
@@ -410,37 +464,38 @@ export class Autocomplete extends Component {
+	withSpokenMessages,
 	withInstanceId,
-	withFocusOutside,
+	withFocusOutside, // this MUST be the innermost HOC as it calls handleFocusOutside
```
Alternatively we could consider incorporating something like `hoist-non-react-statics` so we don't depend on a particular order for composing higher-order components.

https://www.npmjs.com/package/hoist-non-react-statics
